### PR TITLE
[sui-verifier] Added support for characteristic types

### PR DESCRIPTION
**Pull request by awelc**
This PR is the first step towards implementing so called one-time-witness in the form of so called characteristic type. The idea is to guarantee that a given [witness type](http://examples.sui.io/patterns/witness.html) can be used only once.

An example use is to guarantee that there can be only one instance of a treasure capability for a given coin (a "permission-less coin)- treasure capability uses a witness type to [create a new coin](http://examples.sui.io/samples/coin.html).

This is realized by having a very specific type (optionally) defined in a module, and guarantee that this type can be instantiated only once and the instance cannot be stored or copied. This type must fulfill the following properties which are checked in a verifier pass implemented in this PR:
- it has only one ability: drop
- it has only one arbitrarily named field of type boolean (since Move structs cannot be empty)
- its definition does not involve type parameters
- its only instance in existence is passed as an argument to the module initializer
- it is never instantiated anywhere in its defining module

A followup to this PR will be another one that implements a native function checking if a given type is a characteristic type and providing an example of how all this can be used to implement a permission-less coin.

### CODE REVIEW

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:4]: Bit of a naming nit, I think we are going with "one-time-witness type" for anything user facing. As it sort of describes what it is supposed to be used for. 

So a general s/characteristic/one-time-witness/

Comment (tnowacki) [reply crates/sui-verifier/src/char_type_verifier.rs:4]: With or without the hyphens, doesn't matter too much

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:54]: Do you need both of these here? Isn't it just `mod_name.to_ascii_uppercase() == struct_name`?

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:64]: All this would be a bit cleaner with `struct_defs.iter().find`

File: crates/sui-verifier/src/entry_points_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/entry_points_verifier.rs:150]: Nit, could you move this comment next to the code with the checks

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:101]: `AbilitySet::Empty | Ability::Drop` should work

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:103]: You can then do `abilities == drop_set`

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:121]: `signature.0` to avoid the `TypeSignature(...)` wrapper

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [reply crates/sui-verifier/src/char_type_verifier.rs:4]: And open to discussion here :) I just think calling this thing a one-time witness everywhere might help people understand what is going on and why they want to use it (assuming they understand witnesses)

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:69]: nit, some naming here like `witness_name, witness_handle` would make this a bit more readable (at least to me)

I was very confused when looking at the other functions when I just saw `struct_name` and `struct_handle`

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:206]: nit, just early return if it is None

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:208]: nit, I'd probably do one of these `let packed_struct_def_idx = match instr { Bytecode::Pack(idx) => idx, _ => continue };`

Could also do a `filter_map` if you prefer

File: sui_programmability/examples/fungible_tokens/sources/regulated_coin.move
Comment (tnowacki) [inline sui_programmability/examples/fungible_tokens/sources/regulated_coin.move:102]: amazing lol

**Review by tnowacki — COMMENTED** [review]
Overall looks great!

Mostly just layout/naming nits :) (sorry)

We should have one other test where you do something like
```
module a::m {
  struct N has drop {}
}
module a::n {
  fun init(_witness: N, ctx: &mut TxContext) { ... }
}
```
Where you are using the upper case, but from a different module

**Comment by tnowacki** [comment]
One other thing, and we don't have to decide on it in this PR, but should we optionally allow the witness type to have `store`? I know @damirka wanted it.

And do we want the signature of `is_one_time_witness` to be:
`<T>(_witness: T): bool`
or
`<T>(_witness: &T): bool`

The former _forces_ you to use it correctly, but the second gives you a tiny bit more leeway. However we cannot do the by-ref version with `store` so our options are:

A) optional `store` and `<T>(_witness: T): bool`
B) no store and `<T>(_witness: T): bool`
C) no store and `<T>(_witness: &T): bool`

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (awelc) [reply crates/sui-verifier/src/char_type_verifier.rs:4]: To me a characteristic type was a way to implement one-time witness. For now I have updated the comment to reflect that but I also open for a discussion :-)

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (awelc) [reply crates/sui-verifier/src/char_type_verifier.rs:54]: When I was attempting to write an explanation, I realized that you are right :-)

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (awelc) [reply crates/sui-verifier/src/char_type_verifier.rs:64]: Possibly, but then we'd have to deal with a return value being a vector, explain why it would only have one or zero elements, etc. Overall, I am not sure it's a huge win, but happy to do the rewrite if requested one more time :-)

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (awelc) [reply crates/sui-verifier/src/char_type_verifier.rs:101]: Thank you! I felt that there should be a better way...

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (awelc) [reply crates/sui-verifier/src/char_type_verifier.rs:121]: You explicitly tried to avoid it as if I remember correctly you don't love `.0` syntax :-) Changed now!

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (awelc) [reply crates/sui-verifier/src/char_type_verifier.rs:69]: Well, we are looking for structs that can potentially be one-time witness (characteristic) types but at this point it's not guaranteed that they are (they are still candidates subject to verification). Calling them `witness` here is a bit premature I thought...

File: sui_programmability/examples/fungible_tokens/sources/regulated_coin.move
Comment (awelc) [reply sui_programmability/examples/fungible_tokens/sources/regulated_coin.move:102]: It uses `Abc` instances beyond the initializer, but if @damirka can rewrite this code for `Abc` to be a one-time witness type, then we can revert this!

**Comment by awelc** [comment]
> One other thing, and we don't have to decide on it in this PR, but should we optionally allow the witness type to have `store`? I know @damirka wanted it.
> 
> And do we want the signature of `is_one_time_witness` to be: `<T>(_witness: T): bool` or `<T>(_witness: &T): bool`
> 
> The former _forces_ you to use it correctly, but the second gives you a tiny bit more leeway. However we cannot do the by-ref version with `store` so our options are:
> 
> A) optional `store` and `<T>(_witness: T): bool` B) no store and `<T>(_witness: T): bool` C) no store and `<T>(_witness: &T): bool`

Agreed! I went with the most conservative option for now to avoid this discussion before this PR lands. Will likely do the same when implementing the native method as it's easy to change (though we should decide soon!).

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [reply crates/sui-verifier/src/char_type_verifier.rs:4]: I would agree with that description, but I think it is the _only_ way to implement a one-time witness. At least the only safe way. 
And I've been pushing more for naming in the Sui Framework for things to be named what they are, not how they work. Really, the name should reflect the most important and distinguishing part of the thing. 


That being said, I could see one-time witness being confusing as there are lots of ways to implement one-time witness, though all those other ways would be not as safe. 

Thoughts given this perspective?

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [reply crates/sui-verifier/src/char_type_verifier.rs]: `find` gives the first element that satisfies the predicate, so what you have here is equivalent to `find`, no vectors or anything

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [reply crates/sui-verifier/src/char_type_verifier.rs]: I'm here and there, maybe this is going back to my love for my little `sp!()` macro in the source language? I do tend to dislike field access instead of pattern matching

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [reply crates/sui-verifier/src/char_type_verifier.rs:72]: I think once you have the thing that is the `ascii_uppercase` of the module name, that is the characteristic/witness type. After that, you are ensuring it satisfies the necessary properties. 
But I hear you. At the very least, please some name other than `struct_name`, since that could be any struct! Not just your "candidate"

File: sui_programmability/examples/fungible_tokens/sources/regulated_coin.move
Comment (tnowacki) [reply sui_programmability/examples/fungible_tokens/sources/regulated_coin.move:102]: No this is fine, I just found it funny that we had to change it

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:204]: canddiate_

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:167]: candidate_

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:146]: candidate_

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:101]: candidate_

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:170]: if you want to be fancy, `matches(tok, SignatureToken::Struct(idx) if view.struct_handle_at(*idx) == struct_handle)

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:221]: is this what rustfmt gave? Is the string maybe too long to not format this?

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [inline crates/sui-verifier/src/char_type_verifier.rs:72]: Maybe someday we should just move the `init` verification over here as it is a bit strange to have it split. But probably fine for now

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [reply crates/sui-verifier/src/char_type_verifier.rs]: I didn't realize the `Result` in here. Really I think you need `try_find`, which is still in nightly

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (tnowacki) [reply crates/sui-verifier/src/char_type_verifier.rs:4]: Fine with landing the naming as is. I will be curious what @sblackshear thinks, as he came up with the initial name

File: crates/sui-verifier/src/char_type_verifier.rs
Comment (awelc) [reply crates/sui-verifier/src/char_type_verifier.rs:170]: It is fancy indeed!

**Review by tnowacki — APPROVED** [review]
Any other naming discussions can happen after this lands. Thanks for the changes!

**Pull request merged**
